File: docs/Cours/code/chemin.ml
type mat_adj = (bool array) array

let rec chemin (adj : mat_adj) i j len =
  if len = 0 then
       i = j
  else let adj_i = adj.(i) in
       Array.exists (fun k -> adj_i.(k) && chemin adj k j (len-1))
         (Array.init (Array.length adj_i) (fun x -> x) )

File: docs/Cours/code/chemin.c
#include <stdbool.h>

struct mat_adj {
  int n;
  bool **adj;
};

/** M : matrice d'adjacence d'un graphe
 * i, j : indices sommets du graphe
 * len : entier >= 0
 * sortie : vrai si et seulement s'il existe un chemin de i à j
 * de longueur len dans le graphe */
bool chemin(mat_adj M, int i, int j, int len){
  if (len == 0){
    return i == j;
  }
  for(int k=0; k<M.n; k=k+1){
    if (M.adj[i][k] && chemin(M, k, j, len-1)){
      return true;
    }
  }
  return false;
}
